File: LeetCode_Algorithm/98(Validate Binary Search Tree)/98.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from typing import Optional

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        rootNode = self.makeTree(root, 0)
        import math
        stack = [(rootNode, -math.inf, math.inf)]
        # preorder 57ms 69%
        while stack:
            node, lower, upper = stack.pop()
            if not node:
                continue
            if node.val <= lower or node.val >= upper:
                return False
            stack.append((node.right, node.val, upper))
            stack.append((node.left, lower, node.val))
        return True


    def makeTree(self, root, idx):
        rootNode = TreeNode(root[idx])
        if 2*idx+1 < len(root) and root[2*idx+1] != None: rootNode.left = self.makeTree(root,2*idx+1)
        if 2*idx+2 < len(root) and root[2*idx+2] != None: rootNode.right = self.makeTree(root,2*idx+2)
        return rootNode


# print(Solution().isValidBST([2,1,3]))
print(Solution().isValidBST([0,None,-1]))
# print(Solution().isValidBST([5,1,4,None,None,3,6]))

File: LeetCode_Algorithm/98(Validate Binary Search Tree)/98.js
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}
function makeTree(root, idx) {
  rootNode = new TreeNode(root[idx])
  if (2*idx+1 < root.length && root[2*idx+1] != undefined) {
    makeTree(root, 2*idx+1)
  }
  if (2*idx+2 < root.length && root[2*idx+2] != undefined) {
    makeTree(root, 2*idx+2)
  }
  return rootNode
}
var isValidBST = function(root) {
  rootNode = makeTree(root, 0)  
  stack = [[rootNode, -Infinity, Infinity]]
  // preorder O(n) 146ms 8%
  while (stack.length > 0) {
    temp = stack.pop()
    node = temp[0]; lower = temp[1]; upper = temp[2]
    if (node.val <= lower || upper <= node.val) {
      return false
    }
    if (node.right != undefined) {
      stack.push([node.right, node.val, upper])
    }
    if (node.left != undefined) {
      stack.push([node.left, lower, node.val])
    }
  }
  return true
};


console.log(isValidBST([2,1,3]))